File: algorithms/code/leetcode/lc002_add_two_numbers/LC002AddTwoNumbers.java
package leetcode.lc002_add_two_numbers;

/**
 * https://leetcode.com/problems/add-two-numbers/
 * #medium
 */
public final class LC002AddTwoNumbers {
    public static class ListNode {
        int val;
        ListNode next;

        ListNode(final int x) {
            val = x;
        }
    }

    public ListNode addTwoNumbers(final ListNode l1, final ListNode l2) {
        ListNode n1 = l1;
        ListNode n2 = l2;
        ListNode handle = new ListNode(0);
        handle.next = null;
        ListNode tail = handle;
        int carry = 0;
        int value1;
        int value2;
        int sum;
        while (n1 != null || n2 != null || carry != 0) {
            if (n1 == null) {
                value1 = 0;
            } else {
                value1 = n1.val;
                n1 = n1.next;
            }
            if (n2 == null) {
                value2 = 0;
            } else {
                value2 = n2.val;
                n2 = n2.next;
            }
            sum = carry + value1 + value2;
            tail.next = new ListNode(sum % 10);
            tail.next.next = null;
            tail = tail.next;
            carry = sum / 10;
        }
        return handle.next;
    }
}

File: algorithms/code/leetcode/lc002_add_two_numbers/lc002_add_two_numbers.go
// Package lc002 implements https://leetcode.com/problems/add-two-numbers/
package lc002

// ListNode defines a singly-linked list
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1, l2 *ListNode) *ListNode {
	handle := &ListNode{}
	tail := handle
	carry := 0
	for l1 != nil || l2 != nil || carry != 0 {
		value1, value2 := 0, 0
		if l1 != nil {
			value1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			value2 = l2.Val
			l2 = l2.Next
		}
		sum := carry + value1 + value2
		tail.Next = &ListNode{Val: sum % 10}
		tail = tail.Next
		carry = sum / 10
	}
	return handle.Next
}
